projects/worker-ccai/worker-ccai.js

```diff
@@ -0,0 +1,174 @@
+const fs = require('fs').promises;
+const FormData = require('form-data');
+
+const DEFAULT_ANALYZER_ID = "Feature:image-color-histogram:Service-e952f4acd7c2425199b476a2eb459635";
+const DEFAULT_CCAI_ENDPOINT = "https://sensei.adobe.io/services/v1/predict";
+
+/**
+ * @typedef {Object} Color
+ * @property {String} name Color name
+ * @property {Number} percentage Coverage percentage (0.0-1.0)
+ * @property {Number} red Red channel (0-255)
+ * @property {Number} green Green channel (0-255)
+ * @property {Number} blue Blue channel (0-255) 
+ */
+
+/**
+ * Parse the color feature response
+ * 
+ * @param {*} response JSON response from Content and Commerce AI service
+ * @returns {Color[]} Color features returned by the service
+ */
+function parseColors(response) {
+    const colors = [];
+    for (const cas of response.cas_responses) {
+        if (cas.status === 200 && cas.result.response_type === "feature") {
+            for (const r of cas.result.response) {
+                if (r.feature_name === "color") {
+                    for (const value of r.feature_value) {
+                        const [ name, percentage, red, green, blue ] = value.feature_value.split(",");
+                        colors.push({
+                            name,
+                            percentage,
+                            red,
+                            green,
+                            blue
+                        });    
+                    }
+                }
+            }
+        }
+    }
+    return colors;
+}
+
+/**
+ * Sort colors, high coverage to low coverage
+ * 
+ * @param {Color[]} colors Color features
+ * @return {Color[]} Color features sorted by percentage (high to low)
+ */
+function sortColors(colors) {
+    colors.sort((a, b) => b.percentage - a.percentage);
+}
+
+/**
+ * Convert a percentage to a string
+ * 
+ * @param {Color} color Color feature
+ * @returns Percentage as a string, e.g. 59%
+ */
+function toPercentageString(color) {
+    return `${Math.round(color.percentage * 100.0)}%`;
+}
+
+/**
+ * Convert a color feature to a web color
+ * 
+ * @param {Color} color Color feature
+ * @returns Web color, e.g. `#a909fe`
+ */
+function toWebColor(color) {
+    const arr = [color.red, color.green, color.blue];
+    return `#${Buffer.from(arr).toString('hex')}`;
+}
+
+exports.main = worker(async (source, rendition, params) => {
+    // Acquire end point and analyzer
+    const analyzer_id = rendition.instructions.ANALYZER_ID || DEFAULT_ANALYZER_ID;
+    const endpoint = rendition.instructions.CCAI_ENDPOINT || DEFAULT_CCAI_ENDPOINT;
+    console.log("Using analyzer:", analyzer_id);
+    console.log("Using endpoint:", endpoint);
+
+    // Make sure that the source file is not empty
+    const stats = await fs.stat(source.path);
+    if (stats.size === 0) {
+        throw new SourceCorruptError('source file is empty');
+    }
+
+    // Build parameters to send to Sensei service
+    const ext = path.extname(source.path);
```

Comment:
Does the color space of the source already must be RGB? 
FOr instance, what happens with images having CMYK as color profile?

Reply:
It probably fails, I haven't tried it but the C&C API specifies that it only returns RGB.

Reply:
Maybe we could do an example of color space conversion in this worker (using imagemagick, since it's in the base runtime image). Wdyt?

Reply:
IIUC this example worker just supports whatever the CC AI API supports and not much we can do about it... It should document/link to that API documentation, but we don't have to test all that here.

A few simple test cases would be fine. In the end it's an example worker, not a production worker.

BTW, its output is metadata, not an image, so there is no image post processing.

Reply:
Actually this is the question I wanted to raise (in the end): we have *postprocessing*, should we also offer *pre-processing* to cover such cases?

Reply:
Generally we should, AI/ML technology doesn't usually require the full image return accurate information. For example, in this case sending the service a lower resolution version of the image wouldn't change what the most prevalent colors are and their coverage percentages.